clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```diff
@@ -284,8 +284,31 @@ public synchronized void seekToEnd(Collection<TopicPartition> partitions) {
             subscriptions.needOffsetReset(tp, OffsetResetStrategy.LATEST);
     }
 
-    public synchronized void updateEndOffsets(Map<TopicPartition, Long> newOffsets) {
-        endOffsets.putAll(newOffsets);
+    // needed for cases where you make a second call to endOffsets
+    public synchronized void addEndOffsets(final Map<TopicPartition, Long> newOffsets) {
+        innerUpdateEndOffsets(newOffsets, false);
+    }
+
+    public synchronized void updateEndOffsets(final Map<TopicPartition, Long> newOffsets) {
+        innerUpdateEndOffsets(newOffsets, true);
+    }
+
+    private void innerUpdateEndOffsets(final Map<TopicPartition, Long> newOffsets,
+                                       final boolean replace) {
+
+        for (final Map.Entry<TopicPartition, Long> entry : newOffsets.entrySet()) {
+            List<Long> offsets = endOffsets.get(entry.getKey());
+            if (offsets == null) {
+                offsets = new ArrayList<>();
+            }
+            if (replace) {
+                offsets = new ArrayList<>();
```

Comment:
Nit: thinks it's better to include this in L301 above: `if(replace || offsets == null)`